services/victoriametrics/victoriametrics.go

```diff
@@ -0,0 +1,362 @@
+
+// Run runs VictoriaMetrics configuration update loop until ctx is canceled.
+func (svc *VictoriaMetrics) Run(ctx context.Context) {
+	if !svc.enabled {
+		return
+	}
+	svc.l.Info("Starting...")
+	defer svc.l.Info("Done.")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-svc.sema:
+			// batch several update requests together by delaying the first one
+			sleepCtx, sleepCancel := context.WithTimeout(ctx, updateBatchDelay)
+			<-sleepCtx.Done()
+			sleepCancel()
+
+			if ctx.Err() != nil {
+				return
+			}
+
+			if err := svc.updateConfiguration(ctx); err != nil {
+				svc.l.Errorf("Failed to update configuration, will retry: %+v.", err)
+				svc.RequestConfigurationUpdate()
+			}
+		}
+	}
+}
+
+// updateConfiguration updates Prometheus configuration.
+func (svc *VictoriaMetrics) updateConfiguration(ctx context.Context) error {
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > time.Second {
+			svc.l.Warnf("updateConfiguration took %s.", dur)
+		}
+	}()
+
+	cfg, err := svc.marshalConfig()
+	if err != nil {
+		return err
+	}
+
+	return svc.configAndReload(ctx, cfg)
+}
+
+// RequestConfigurationUpdate requests VictoriaMetrics configuration update.
+func (svc *VictoriaMetrics) RequestConfigurationUpdate() {
+	if !svc.enabled {
+		return
+	}
+	select {
+	case svc.sema <- struct{}{}:
+		ctx, cancel := context.WithTimeout(context.Background(), configurationUpdateTimeout)
+		defer cancel()
+		err := svc.updateConfiguration(ctx)
+		if err != nil {
+			svc.l.WithError(err).Errorf("cannot reload configuration")
+		}
+
+	default:
+	}
+}
+
+// IsReady verifies that VictoriaMetrics works.
+func (svc *VictoriaMetrics) IsReady(ctx context.Context) error {
+	if !svc.enabled {
+		return nil
+	}
+	// check VictoriaMetrics /health API
+	u := *svc.baseURL
+	u.Path = path.Join(u.Path, "health")
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	resp, err := svc.client.Do(req)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+	b, err := ioutil.ReadAll(resp.Body)
+	svc.l.Debugf("VictoriaMetrics: %s", b)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return errors.Errorf("expected 200, got %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+// reload asks VictoriaMetrics to reload configuration.
+func (svc *VictoriaMetrics) reload(ctx context.Context) error {
+	u := *svc.baseURL
+	u.Path = path.Join(u.Path, "-", "reload")
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	resp, err := svc.client.Do(req)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return errors.Errorf("%d: %s", resp.StatusCode, b)
```

Comment:
Could it be a shared function in Prometheus package?

Reply:
Well, we are going to remove prometheus package in the next release. It will allow us to make less changes.

Reply:
```
// ScrapeTimeout - wraps scrapeTimeout and makes it public for victoriametrics package.
 func ScrapeTimeout(interval time.Duration) config.Duration
```

So we did that to make less changes, but copy&paste code the to make less changes? Why we don't expose that function like we did with ScrapeTimeout?

Reply:
probably, this function can be unified with `vmalert`. wdyt?

Reply:
smthng like:
```go
func reloadConfig( ctx, url, wantStatusCode) error
```

Reply:
Sounds good.

Also, that's something I would not insist on if it would take a lot of time

Reply:
As far as i remember, we discussed it - changes to prometheus config generation shouldn't impact other people work and be minimal. That's why neither ScrapeTimeout nor ScrapeConfig generation were copy-pasted.
 
 And also, wrapping isReady and reload funcs for prometheus package leads to additional changes.

 I've tried to rewrite reloadConfig func and i don't like how it looks.. Lets refactor it next time.